src/main/java/picard/illumina/IlluminaBasecallsToFastq.java

```diff
@@ -361,203 +370,279 @@ private void populateWritersFromMultiplexParams() {
-            this.templateWriters = templateWriters;
-            this.sampleBarcodeWriters = sampleBarcodeWriters;
-            this.molecularBarcodeWriters = molecularBarcodeWriters;
-        }
-
+    private static final class NoOpClusterConverter implements BasecallsConverter.ClusterDataConverter<ClusterData> {
         @Override
-        public void write(final FastqRecordsForCluster records) {
-            write(templateWriters, records.templateRecords);
-            write(sampleBarcodeWriters, records.sampleBarcodeRecords);
-            write(molecularBarcodeWriters, records.molecularBarcodeRecords);
+        public ClusterData convertClusterToOutputRecord(ClusterData cluster) {
+            return cluster;
         }
+    }
 
-        private void write(final FastqWriter[] writers, final FastqRecord[] records) {
-            for (int i = 0; i < writers.length; ++i) {
-                writers[i].write(records[i]);
-            }
-        }
+    private final class AsyncClusterWriter extends AbstractAsyncWriter<ClusterData> implements BasecallsConverter.ConvertedClusterDataWriter<ClusterData>  {
+        private final ClusterWriter writer;
 
-        @Override
-        public void close() {
-            for (final FastqWriter writer : templateWriters) {
-                writer.close();
-            }
-            for (final FastqWriter writer : sampleBarcodeWriters) {
-                writer.close();
-            }
-            for (final FastqWriter writer : molecularBarcodeWriters) {
-                writer.close();
-            }
+        public AsyncClusterWriter(final ClusterWriter out, final int queueSize) {
+            super(queueSize);
+            this.writer = out;
         }
-    }
 
-    /**
-     * Contains the results of transforming one cluster into the record(s) to be written to output file(s).
-     */
-    static class FastqRecordsForCluster {
-        // These are accessed directly by converter and writer rather than through getters and setters.
-        final FastqRecord[] templateRecords;
-        final FastqRecord[] sampleBarcodeRecords;
-        final FastqRecord[] molecularBarcodeRecords;
-
-        FastqRecordsForCluster(final int numTemplates, final int numSampleBarcodes, final int numMolecularBarcodes) {
-            templateRecords = new FastqRecord[numTemplates];
-            sampleBarcodeRecords = new FastqRecord[numSampleBarcodes];
-            molecularBarcodeRecords = new FastqRecord[numMolecularBarcodes];
-        }
+        @Override protected String getThreadNamePrefix() { return "FastqWriterThread-"; }
+        @Override protected void synchronouslyWrite(final ClusterData item) { this.writer.write(item); }
+        @Override protected void synchronouslyClose() { this.writer.close(); }
     }
-
     /**
-     * Passed to IlluminaBaseCallsConverter to do the conversion from input format to output format.
+     * An optimized writer for writing ClusterData directly to a set of Fastq files.
```

Comment:
This is basically pushing the compute of decoding cluster data to the writer instead of the reader which makes it so that we can better optimize threading across all writers. This will make a bigger impact when the new AsyncWriter that is being worked on in htsjdk is wired in.